examples/hello_debugger.cpp

```diff
@@ -84,24 +119,45 @@ class Debugger {
   std::mutex mutex;
   int line = 1;
   std::unordered_set<int> breakpoints;
+  std::atomic<bool> terminate{false};
+  std::atomic<bool> running{true};
+  ::Event canRun;
 };
 
 Debugger::Debugger(const EventHandler& onEvent) : onEvent(onEvent) {}
 
 void Debugger::run() {
-  std::unique_lock<std::mutex> lock(mutex);
-  for (int i = 0; i < numSourceLines; i++) {
-    auto l = ((line + i) % numSourceLines) + 1;
-    if (breakpoints.count(l)) {
-      line = l;
+  while (!terminate) {
+    // If not running, wait
+    if (!running) {
+      canRun.wait();
+    }
+
+    // Go to next line
+    line = (line % numSourceLines) + 1;
+
+    // Check for breakpoints
+    std::unique_lock<std::mutex> lock(mutex);
+    if (breakpoints.count(line)) {
       lock.unlock();
+      running = false;
       onEvent(Event::BreakpointHit);
-      return;
     }
   }
 }
 
+void Debugger::stop() {
+  terminate = true;
+  resume();
+}
+
+void Debugger::resume() {
+  running = true;
```

Comment:
Although you're using an atomic, this is racy. `running` can be set to `false` at line `143`.

Reply:
I'm not sure I see the problem here. If a separate thread were to set running to false at line 143 at the same time as another thread sets it to true here, things still remain consistent. In effect, hitting a breakpoint and pressing continue at the same time would mean one of them would win. In reality, the UI wouldn't let you press continue unless you were already paused, so it should be fine.

Maybe I'm not seeing it, though. What would you suggest here?

Reply:
My concern is that whenever I see the usage of `std::atomic<bool>` as signals there's usually a race condition. Here there is, but it looks like it'll recover.
However, this is example code, and I'm not sure we should be demonstrating patterns that can hide raciness.

Ideally in this sort of set up, I like to use a thread-safe queue of events for controlling the worker, and a query API for current state. This seems overkill for what we have here. Instead, we can adapt the `Event` a bit, and still keep things synchronized with a mutex lock: https://github.com/ben-clayton/cppdap/commit/3027fc089769813dc5e8913f85becbed14a977de

What do you think?

Reply:
>> Instead, we can adapt the Event a bit, and still keep things synchronized with a mutex lock: ben-clayton@3027fc0 
>> What do you think?

Yes, I like this a lot more. I think it 'test' should be called 'test_and_reset' to make it clear that it resets the fire state, but yeah, your solution of using two events is much better. In fact, originally I had wanted to query to state of the "canRun" event, but saw that it didn't have that, and figured an atomic<bool> would do the same; but it's less clean, as you pointed out.

Reply:
`test_and_reset` - maybe, it's a bit verbose though. Note that `wait()` also resets. I did update to comments on both. FWIW, a `test()` without resetting the signal state is also a bad idea for all the reasons I described above.

The sleep is probably quite excessive. Feel free to reduce it.

Yes, I gave it a quick test. Seemed to do the trick.